service/frontend/adminHandler.go

```diff
@@ -529,9 +529,87 @@ func (adh *AdminHandler) CloseShard(ctx context.Context, request *adminservice.C
 	return &adminservice.CloseShardResponse{}, err
 }
 
-// TODO: consolidate into one ListShardTasks API and the range in request should be [inclusive, exclusive) for all task categories
-// Currently for transfer, replication, and visibility tasks the range in the request is (exclusive, inclusive], so add 1 when
-// calling persistence layer.
+func (adh *AdminHandler) ListHistoryTasks(
+	ctx context.Context,
+	request *adminservice.ListHistoryTasksRequest,
+) (_ *adminservice.ListHistoryTasksResponse, retError error) {
+	defer log.CapturePanic(adh.logger, &retError)
+
+	scope, sw := adh.startRequestProfile(metrics.AdminListHistoryTasksScope)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+	taskRange := request.GetTaskRange()
+	if taskRange == nil {
+		return nil, adh.error(errTaskRangeNotSet, scope)
+	}
+
+	var taskCategory tasks.Category
+	switch request.Category {
+	case enumsspb.TASK_CATEGORY_UNSPECIFIED:
+		return nil, adh.error(errTaskCategoryNotSet, scope)
+	case enumsspb.TASK_CATEGORY_TRANSFER:
+		taskCategory = tasks.CategoryTransfer
+	case enumsspb.TASK_CATEGORY_TIMER:
+		taskCategory = tasks.CategoryTimer
+	case enumsspb.TASK_CATEGORY_REPLICATION:
+		taskCategory = tasks.CategoryReplication
+	case enumsspb.TASK_CATEGORY_VISIBILITY:
+		taskCategory = tasks.CategoryVisibility
+	default:
+		var ok bool
+		taskCategory, ok = tasks.GetCategoryByID(int32(request.Category))
+		if !ok {
+			return nil, adh.error(&serviceerror.InvalidArgument{
+				Message: fmt.Sprintf("unknown task category ID: %v", int32(request.Category)),
+			}, scope)
+		}
```

Comment:
can we just use this block and remove the switch clause?

Reply:
Currently we can't, the order of those categories in proto is not the same as the order defined in history service. 

(I can change them to match though, since the change is not released yet.)

Reply:
https://github.com/temporalio/temporal/pull/2586
Removed the switch clause.